File: impl/go/internal/engine/effects_test.go
package engine

import (
	"testing"

	"github.com/zaibatsu/zaibatsu-go/internal/domain"
)

func allFacesDefense() []domain.DefenseDie {
	out := make([]domain.DefenseDie, 6)
	for i := range out {
		out[i] = domain.DefenseDie{Value: i + 1}
	}
	return out
}

func effectData(t *testing.T) *domain.GameData {
	t.Helper()
	gd := loadOrSkip(t)
	cloned := *gd
	cloned.Blocks = append([]domain.Block{}, gd.Blocks...)
	cloned.Pawns = append([]domain.Pawn{}, gd.Pawns...)
	for i := range cloned.Blocks {
		if cloned.Blocks[i].ID == "data-haven" {
			cloned.Blocks[i].Effects = domain.BlockEffects{
				InCybernet:   &domain.BlockEffect{Kind: "area-attack", Amount: 1},
				UnderControl: &domain.BlockEffect{Kind: "area-attack", Amount: 1},
			}
		}
	}
	for i := range cloned.Pawns {
		if cloned.Pawns[i].ID == "speedrunner-yellow" {
			cloned.Pawns[i].Defense = allFacesDefense()
		}
	}
	return &cloned
}

func TestApplyBlockEffectAreaAttack(t *testing.T) {
	gd := effectData(t)
	s, _ := NewGame(Config{Data: gd, PlayerNames: []string{"A", "B"}, Seed: 1})
	origin := domain.Coord{Q: 0, R: 0}
	if _, err := PlaceBlock(s, origin, 0, gd, "data-haven", rotFacing(t, gd, "data-haven", 0)); err != nil {
		t.Fatalf("place: %v", err)
	}
	coord := origin.Neighbor(0)
	s.Cybernet.Pawns = []*domain.PawnOnBoard{}
	s.Cybernet.PlacePawn(&domain.PawnOnBoard{PawnID: "speedrunner-red", OwnerID: "p1", Coord: coord, SpaceID: "a"})
	s.Cybernet.PlacePawn(&domain.PawnOnBoard{PawnID: "speedrunner-yellow", OwnerID: "p2", Coord: coord, SpaceID: "a"})

        res := ApplyBlockEffect(s, gd, coord, &domain.BlockEffect{Kind: "area-attack", Amount: 1})
        if res == nil {
		t.Fatal("expected typed area-attack effect result")
	}
        if res.Kind != "area-attack" || res.Result == nil {
                t.Fatalf("unexpected effect result: %#v", res)
        }
        if len(res.Result.Targets) != 2 {
                t.Fatalf("expected 2 targets, got %d", len(res.Result.Targets))
	}
	if s.Cybernet.PawnByID("speedrunner-yellow") != nil {
		t.Fatal("speedrunner-yellow should be eliminated by the block effect")
	}
}

func TestApplyBlockEffectForTriggerUsesPlacedBlockEffect(t *testing.T) {
	gd := effectData(t)
	s, _ := NewGame(Config{Data: gd, PlayerNames: []string{"A", "B"}, Seed: 1})
	origin := domain.Coord{Q: 0, R: 0}
	if _, err := PlaceBlock(s, origin, 0, gd, "data-haven", rotFacing(t, gd, "data-haven", 0)); err != nil {
		t.Fatalf("place: %v", err)
	}
	coord := origin.Neighbor(0)
	s.Cybernet.Pawns = []*domain.PawnOnBoard{}
	s.Cybernet.PlacePawn(&domain.PawnOnBoard{PawnID: "speedrunner-red", OwnerID: "p1", Coord: coord, SpaceID: "a"})
	s.Cybernet.PlacePawn(&domain.PawnOnBoard{PawnID: "speedrunner-yellow", OwnerID: "p2", Coord: coord, SpaceID: "a"})

	res := ApplyBlockEffectForTrigger(s, gd, coord, BlockEffectInCybernet)
        if res == nil {
		t.Fatal("expected trigger lookup to return an effect result")
	}
        if res.Kind != "area-attack" || res.Result == nil {
                t.Fatalf("unexpected effect result: %#v", res)
        }
	if s.Cybernet.PawnByID("speedrunner-yellow") != nil {
		t.Fatal("speedrunner-yellow should be eliminated by the placed block's trigger")
	}
}

func TestApplyBlockEffectForTriggerPlacePawnOnControl(t *testing.T) {
        gd := loadOrSkip(t)
        cloned := *gd
        cloned.Blocks = append([]domain.Block{}, gd.Blocks...)
        for i := range cloned.Blocks {
                if cloned.Blocks[i].ID == "idoru" {
                        cloned.Blocks[i].IceFaces = []int{1, 2, 3, 4, 5, 6}
                }
        }
        gd = &cloned
        s, _ := NewGame(Config{Data: gd, PlayerNames: []string{"A", "B"}, Seed: 1})
        origin := domain.Coord{Q: 0, R: 0}
        if _, err := PlaceBlock(s, origin, 0, gd, "idoru", rotFacing(t, gd, "idoru", 0)); err != nil {
                t.Fatalf("place: %v", err)
        }
        coord := origin.Neighbor(0)
        s.Cybernet.Pawns = []*domain.PawnOnBoard{}
        s.Cybernet.PlacePawn(&domain.PawnOnBoard{PawnID: "speedrunner-red", OwnerID: "p1", Coord: coord, SpaceID: "a"})

        res, err := IcebreakBlock(s, gd, "speedrunner-red", coord, 0)
        if err != nil {
                t.Fatalf("IcebreakBlock: %v", err)
        }
        if !res.Success {
                t.Fatal("expected IcebreakBlock to succeed")
        }
        placed := s.Cybernet.PawnByID("idoru")
        if placed == nil {
                t.Fatal("expected Idoru to be placed by the block effect")
        }
        if placed.OwnerID != "p1" || placed.Coord != coord {
                t.Fatalf("idoru = %#v, want owner p1 at %v", placed, coord)
        }
}

File: impl/ts/test/effects.test.ts
import { describe, expect, test } from "bun:test";
import { loadDefault } from "../src/data/index.ts";
import { blockById, type DefenseDie, type GameData } from "../src/domain/types.ts";
import { neighbor, opposite } from "../src/domain/hex.ts";
import { newGame, placeBlock, applyBlockEffectForTrigger, applyBlockEffect, icebreakBlock } from "../src/engine/index.ts";

const ORIGIN = { q: 0, r: 0 };

function allFacesDefense(): DefenseDie[] {
  return [1, 2, 3, 4, 5, 6].map((value) => ({ value, shielded: false }));
}

function effectData(): GameData {
  const d = structuredClone(loadDefault("speedrunners"));
  d.pawns = d.pawns.map((pawn) =>
    pawn.id === "speedrunner-yellow" ? { ...pawn, defense: allFacesDefense() } : pawn
  );
  d.blocks = d.blocks.map((block) =>
    block.id === "data-haven"
      ? {
        ...block,
        effects: {
          inCybernet: { kind: "area-attack", amount: 1 },
          underControl: { kind: "area-attack", amount: 1 },
        },
      }
      : block
  );
  return d;
}

function rotFacing(data: GameData, blockId: string, dir: number): number {
  const b = blockById(data, blockId)!;
  for (let rot = 0; rot < 6; rot++) {
    const local = (((opposite(dir) - rot) % 6) + 6) % 6;
    if (b.edges![local]) return rot;
  }
  throw new Error(`block ${blockId} cannot face direction ${dir}`);
}

describe("block effects", () => {
  test("typed area-attack effects resolve against every pawn on the block", () => {
    const d = effectData();
    const s = newGame({ data: d, playerNames: ["A", "B"], seed: 1 });
    placeBlock(s, ORIGIN, 0, d, "data-haven", rotFacing(d, "data-haven", 0));
    const coord = neighbor(ORIGIN, 0);
    s.cybernet.pawns = [];
    s.cybernet.placePawn({ pawnId: "speedrunner-red", ownerId: "p1", coord, spaceId: "a" });
    s.cybernet.placePawn({ pawnId: "speedrunner-yellow", ownerId: "p2", coord, spaceId: "a" });

    const res = applyBlockEffect(s, d, coord, { kind: "area-attack", amount: 1 })!;

    expect(res.kind).toBe("area-attack");
    expect(res.result.targets.map((target) => target.targetPawnId)).toEqual(["speedrunner-red", "speedrunner-yellow"]);
    expect(s.cybernet.pawnById("speedrunner-yellow")).toBeUndefined();
  });

  test("trigger lookup reads the placed block's typed effect", () => {
    const d = effectData();
    const s = newGame({ data: d, playerNames: ["A", "B"], seed: 1 });
    placeBlock(s, ORIGIN, 0, d, "data-haven", rotFacing(d, "data-haven", 0));
    const coord = neighbor(ORIGIN, 0);
    s.cybernet.pawns = [];
    s.cybernet.placePawn({ pawnId: "speedrunner-red", ownerId: "p1", coord, spaceId: "a" });
    s.cybernet.placePawn({ pawnId: "speedrunner-yellow", ownerId: "p2", coord, spaceId: "a" });

    const res = applyBlockEffectForTrigger(s, d, coord, "inCybernet");

    expect(res?.kind).toBe("area-attack");
    expect(s.cybernet.pawnById("speedrunner-yellow")).toBeUndefined();
  });

  test("under-control place-pawn effects place the authored pawn on successful Icebreak", () => {
    const d = structuredClone(loadDefault("speedrunners"));
    d.blocks = d.blocks.map((block) => (
      block.id === "idoru" ? { ...block, iceFaces: [1, 2, 3, 4, 5, 6] } : block
    ));
    const s = newGame({ data: d, playerNames: ["A", "B"], seed: 1 });
    placeBlock(s, ORIGIN, 0, d, "idoru", rotFacing(d, "idoru", 0));
    const coord = neighbor(ORIGIN, 0);
    s.cybernet.pawns = [];
    s.cybernet.placePawn({ pawnId: "speedrunner-red", ownerId: "p1", coord, spaceId: "a" });

    const res = icebreakBlock(s, d, "speedrunner-red", coord);

    expect(res.success).toBe(true);
    expect(s.cybernet.at(coord)?.ownerId).toBe("p1");
    expect(s.cybernet.pawnById("idoru")).toEqual(expect.objectContaining({ ownerId: "p1", coord }));
  });
});
